File: trading_system/runtime_config.py
from __future__ import annotations

import os
from pathlib import Path

REPO_ROOT = Path(os.environ.get('SPOT_BOT_REPO_ROOT', Path(__file__).resolve().parents[1]))
BOT_HOME = Path(os.environ.get('HERMES_SPOT_BOT_HOME', os.environ.get('HOME', str(REPO_ROOT))))
DATA_DIR = Path(os.environ.get('AUTO_TRADER_DATA_DIR', str(BOT_HOME / '.trading-data')))
DB_PATH = Path(os.environ.get('AUTO_TRADER_DB_PATH', str(DATA_DIR / 'trading.db')))
BRIDGE_DIR = DATA_DIR / 'telegram-bridge'
QUEUE_FILE = DATA_DIR / 'spot_recovery_manual_review.jsonl'
TELEGRAM_TOKEN_FILE = Path(os.environ.get('TELEGRAM_TOKEN_FILE', str(REPO_ROOT / 'telegram.txt')))
TELEGRAM_TRADE_CHAT_ID = os.environ.get('TELEGRAM_TRADE_CHAT_ID', '2116422114')
JUP_BIN = os.environ.get('JUP_BIN', 'jup')
HELIUS_BIN = os.environ.get('HELIUS_BIN', 'helius')
MONITOR_PATH = Path(os.environ.get('SPOT_MONITOR_PATH', str(REPO_ROOT / 'monitor.mjs')))
SESSIONS_INDEX = REPO_ROOT / '.hermes' / 'sessions' / 'sessions.json'
SESSIONS_DIR = REPO_ROOT / '.hermes' / 'sessions'


def build_path_env() -> str:
    extras = []
    home = os.environ.get('HOME')
    if os.environ.get('JUP_BIN_DIR'):
        extras.append(os.environ['JUP_BIN_DIR'])
    if os.environ.get('HELIUS_BIN_DIR'):
        extras.append(os.environ['HELIUS_BIN_DIR'])
    if home:
        extras.append(str(Path(home) / '.hermes' / 'node' / 'bin'))
        extras.append(str(Path(home) / '.cargo' / 'bin'))
    extras = [p for p in extras if p]
    return ':'.join(extras + [os.environ.get('PATH', '')]) if extras else os.environ.get('PATH', '')

File: runtime-config.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
export const REPO_ROOT = path.dirname(__filename);
const CONFIG_FILE = process.env.SPOT_BOT_CONFIG_FILE || path.join(REPO_ROOT, 'config', 'local.json');

function readConfig() {
  try {
    if (!fs.existsSync(CONFIG_FILE)) return {};
    const raw = JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
    return raw && typeof raw === 'object' ? raw : {};
  } catch {
    return {};
  }
}

const CONFIG = readConfig();
const cfg = (key, fallback) => process.env[key] || CONFIG[key] || fallback;

export const BOT_HOME = cfg('HERMES_SPOT_BOT_HOME', process.env.HOME || REPO_ROOT);
export const DATA_DIR = cfg('AUTO_TRADER_DATA_DIR', path.join(BOT_HOME, '.trading-data'));
export const DB_PATH = cfg('AUTO_TRADER_DB_PATH', path.join(DATA_DIR, 'trading.db'));
export const TRADING_MD = cfg('AUTO_TRADER_TRADING_MD', path.join(REPO_ROOT, 'trading.md'));
export const DB_CLI = cfg('AUTO_TRADER_DB_CLI', path.join(REPO_ROOT, 'trading_system', 'trading_db_cli.py'));
export const JUP_BIN = cfg('JUP_BIN', 'jup');
export const HELIUS_BIN = cfg('HELIUS_BIN', 'helius');

const inferredBinDirs = [
  process.env.JUP_BIN_DIR || CONFIG.JUP_BIN_DIR,
  process.env.HELIUS_BIN_DIR || CONFIG.HELIUS_BIN_DIR,
  process.env.HOME ? path.join(process.env.HOME, '.hermes', 'node', 'bin') : '',
  process.env.HOME ? path.join(process.env.HOME, '.cargo', 'bin') : '',
].filter(Boolean);

export const PATH_ENV = inferredBinDirs.length ? `${inferredBinDirs.join(':')}:${process.env.PATH || ''}` : (process.env.PATH || '');
export const SPOT_BOT_CONFIG_FILE = CONFIG_FILE;
